### ISSUE TITLE
Pause in "on start" does not delay "forever" loops?

### ISSUE THREAD

**Issue by microbit-carlos** [Microsoft/pxt#2385]
microbit.org-ticket: 2187

Aren't the "forever" loops meant to start running after the "on start" blocks have finished?

If we need to initialise something before it can be used (like neopixels or game sprites), the "on start" is usually the preferred place to do that, but if the "forever" loops are declared before the "on start" code, and then the main fibre sleeps, you could have a race condition like in the following example:

![microbit-screenshot](https://user-images.githubusercontent.com/29712657/27773713-e13ff006-5f77-11e7-8689-3b686d53e81d.png)

This program produces the following error and crashes the emulator:

```
Program Error: r.isRef is not a function
```

Hex file with this minimalist example: [issue.zip](https://github.com/Microsoft/pxt/files/1118125/issue.zip)

**Comment by guillaumejenkins** [comment]
All event handlers and forever loops are registered / started before the `on start` block gets executed. Also, the `on start` block, the `forever` loop and all event handlers are executed in parallel, so pausing execution in `on start` does not affect the execution of `forever` or any event handler.

If you need to delay execution in an event handler or in a forever loop until after something happens in `on start`, a good approach is to use a variable:

![image](https://user-images.githubusercontent.com/14299377/27798740-37685c80-5fc7-11e7-8f6c-fb1a5f753452.png)

**Comment by abchatra** [comment]
Forever is a background fiber. It will run as soon as it gets a chance. Pause in the onstart will give that chance to the forever fiber. This is by design.

**Comment by microbit-carlos** [comment]
Thank you very much for your responses. Adding a flag to indicate "readiness" is certainly a workaround, but it cannot be denied that the current implementation is not really intuitive. The user would expect the application to "pause" for an amount of time, and instead other parts of the program run earlier.

The user has no way to know that when a "pause" block is used within "on start", it will automatically trigger processing of the other fibers (in this case the "for ever" blocks). So they would have no way to know that the would need to create a "ready" flag to avoid this issue. On top of that, the exposed error "Program Error: r.isRef is not a function" is cryptic and gives absolutely no indication of where the issue might lie.

I understand that forever being a background fibre is a design decision,  but is it done "by design" that the pause block on the "on start" will trigger them? This behavior would be very difficult to predict unless the user has a very good knowledge of the internal micro:bit runtime implementation.

**Comment by abchatra** [comment]
There are advantages to for transferring the control to waiting fibers. Events like OnShake will be responsive. Micro:bit is a non-preemptive runtime. So unless you have a pause (which is kind of a sleep) control won't be given to other fibers.

**Comment by microbit-carlos** [comment]
Thanks for the follow up @abchatra. I'm just wondering if it's there any advantage to run any fibre before `on start` finishes? Just from a user perspective, I would expect it to run first and finish before anything else starts running. And as it currently stands, other event like the mentioned `onShake` would only work before the `on start` finishes if the user happens to include a `pause` block, which is a big "if".

As a non-preemtive scheduler I understand the need to protect against infinite loops (without a sleep) to block other fibers/events, but right now there is no protection for that use-case, so what is the advantage of declaring the fibres before `on start`?

**Comment by riknoll** [comment]
Declaring the event handlers before the on start protects against infinite loops *with* sleeps. If we declared event handlers after on start then an infinite loop would prevent them from ever being registered (even if that loop had a sleep).

**Comment by microbit-carlos** [comment]
Thanks for the follow up @riknoll. Okay, so it would be helpful to update the documentation to indicate that the `pause` block would cause the other fibres to start executing, as this is not an intuitive behaviour.

**Comment by abchatra** [comment]
Need discussion here as per Jonny

**Comment by DaveAtKitronik** [comment]
I stumbled across this issue in a different and subtle manner recently (see micro:bit ticket 2640). Jonny suggested bringing the discussion here, so:

Some blocks have implicit pauses internally - with no obvious indication - consider:

![in error](https://user-images.githubusercontent.com/25896029/30365733-8ffaea62-9860-11e7-939a-8695f3ac0010.PNG)

![all ok](https://user-images.githubusercontent.com/25896029/30365619-309e009a-9860-11e7-91ed-b7fca5b8f09b.PNG)

I would have assumed that show icon merely turned on some GPIO to light the LEDs - it appears not.

This subtlety creates hidden block order dependencies that do not have a good error message to assist in the debugging. 

My mental model of 'on start' is that it is a bit like a constructor - it sets up the world ready to run, and then the world runs. 
It is currently actually more like:  'first in the queue to execute, but may not complete before other stuff starts.'  This causes the metaphor to be broken - potentially leading to confusion. 

In the micro:bit ticket I proposed a couple of solutions that might be more appropriate to unskilled users. Neither is perfect, so discussion is needed.
One was to only allow a subset of items to be placed into 'on start' - This is complex, and potentially limiting.
Another was to build the 'Start has completed' workaround flag into the forever loop and other event handlers. This probably has subtle implications I am not aware of.

**Comment by abchatra** [comment]
@riknoll 
This is the idea suggested by Jonny:

- Register all the events such as onButtonPressed before the OnStart
- Register basic.forever after onStart

I do like this suggestion. What do you think?

**Comment by riknoll** [comment]
Works for me! I can add a comment attribute that specifies whether an API should appear before or after on start. I think this just applies to forever right now, but maybe `control.runInBackground()` as well. That is, unless we decide to [allow that to be nested](https://github.com/Microsoft/pxt/pull/2660) which is a breaking change (but one I'm personally in favor of).

**Comment by riknoll** [comment]
BTW this is a breaking change. It won't affect 99.9% of scripts, but it is breaking.

**Comment by abchatra** [comment]
Lets go with this fix. Thanks!

**Comment by pelikhan** [comment]
@mmoskal @tballmsft FYI

**Comment by pelikhan** [comment]
Run  in background is another one we should fix. It should be nestable.

**Comment by riknoll** [comment]
Let's move that to a different thread

**Comment by mmoskal** [comment]
I think treating `forever` and `onShake` differently is inconsistent and doesn't really fix the problem. The initial example could have as well used `onShake` and not `forever`, and it would crash. IMHO either we register all events before the `on start` or all after. I think it's safer to do after, but indeed it is a breaking change.

**Comment by riknoll** [comment]
I disagree. `onshake` requires user input on the hardware whereas `forever` runs immediately (or as soon as it can). They have different semantics so it's fair to treat them differently.

**Comment by mmoskal** [comment]
@riknoll then it's even worse - you will non-deterministically get crashes depending on how fast you shake.

In some sense it's fair for us to order it as we feel like in JS, since there is no intrinsic order to the blocks...

**Comment by riknoll** [comment]
I guess that's true, but it's always been the case. The argument can also be made that the wording of the block `forever` vs `onSomeEvent` means the user might not expect them to behave exactly the same when compiled.

**Comment by DaveAtKitronik** [comment]
I would expect - trying to think from a non expert POV - that if I programmed on start it would be the first thing to run, and it would complete before forever - which is the only other 'code' area I write - starts.
I am not sure how a non expert would view hardware driven events (on shake etc). 
I think that it would be reasonable to expect them to 'just work'. 
This code:

![shakybreaky](https://user-images.githubusercontent.com/25896029/30418662-a2b64aae-992b-11e7-9fe4-b109732e3bae.PNG)

Runs fine in the simulator, but occasionally crashes on a real microbit. It can be made to crash by shaking the microbit as you release the reset button - not sure how 'real world' that is. Once crashed the reset button does not bring it back - only a power cycle will.

**Comment by microbit-carlos** [comment]
From my point of view having "events" like on shake or a button press execute before the "on start" finishes it's a valid expectation, as you could easily explain that you are triggering that event on micro:bit before it finishes execution that portion of the code.
However, they way blocks like `forever` and `run in background` are named do no "feel" like events, in this case it is harder to explain the fiber model without introducing other concepts first, and why a pause will cause the other code to execute.

**Comment by abchatra** [comment]
Also anyone coming from Arduino world, expectation is forever starts only after on-start code executed. We differ from there. My vote is to original proposal from @jaustin

1. Register all the user events such as onButtonPressed before the OnStart
2. Register basic.forever and run in background after onStart

**Comment by mmoskal** [comment]
I guess this is fine, since the behavior is crystal-clear in JavaScript.

**Comment by abchatra** [comment]
@riknoll this is ported to v0 right?

**Comment by abchatra** [comment]
fixed

**Comment by lct4246** [comment]
The same problem, Somebody can help me.
I can't resolve this Gold Rush code (http://www.101computing.net/gold-rush/).
Here is the hex file. https://drive.google.com/file/d/0B_9NUWwOyRGISTJBT1pYSzhVSDg/view?usp=sharing

**Comment by pelikhan** [comment]
It seems that the conversion failed on a variable assignment. Try this patched version:

    https://makecode.microbit.org/_TVWKArCuCCFT

**Comment by lct4246** [comment]
Thank you so much.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Ted Chun-Te Lee 
李俊德（你記得）

URL（Google keyword no 1：Ted's Bu　Bu）：https://sites.google.com/site/lct4246/
FB（Google keyword no 1：Ted's BuBu）：https://www.facebook.com/ted.lee.10420321
Blog（Google keyword no 1：泰布布）：http://lct4246.blogspot.tw/

**Comment by ahnniu** [comment]
Hi @abchatra,

> Register all the user events such as onButtonPressed before the OnStart
Register basic.forever and run in background after onStart

I think there will be some issues when "Register all the user events such as onButtonPressed before the OnStart". Let's guess:

- When an object has some events
    + We should ensure that the object variable should be initialised first before the "event register block"
- When an event register block use some variables
    + The variables should be initialised first too

All of them will run to error: "Program Error: r.isRef is not a function".

See another issue:

- [pxt.microbit.org variable initialisation -> control.onEvent problem #1830](https://github.com/Microsoft/pxt/issues/1830)

I'm now implement some class with events. In the current situation, the object initializing is very late and will result "Program Error: r.isRef is not a function".

**Comment by abchatra** [comment]
@ahnniu can you share your program?

**Comment by ahnniu** [comment]
Hi abchatra,

See below.

![microbit-screenshot 1](https://user-images.githubusercontent.com/7505963/36575470-1db88274-1886-11e8-83d5-83a5d81e4d3b.png)


Typescript code:

```javascript
let line: bot.LineFollower = null
line.onEvent(LineFollowerEvent.LeftOut, () => {

})
line = bot.createLineFollower(1)
basic.forever(() => {

})
```

**Comment by abchatra** [comment]
What package are you using?

**Comment by abchatra** [comment]
Easiest way to share the program is to use Share button?

**Comment by ahnniu** [comment]
Hi @abchatra ,

The package is developed by me and I'm now debugging it.  I can paste some critical code of the package here:

```typescript
enum LineFollowerEvent {
    /**
     * Left side goes out of black line
     */
    //% block="left out"
    LeftOut = 1,
    /**
     * Left side goes inside of black line
     */
    //% block="left in"
    LeftIn = 2,
    /**
     * Right side goes out of black line
     */
    //% block="right out"
    RightOut = 3,
    /**
     * Right side goes inside of black line
     */
    //% block="right in"
    RightIn = 4,
}

namespace bot {

    //% shim=bot::lowLevelLineFollowerConstruct
    function lowLevelLineFollowerConstruct(index: number): number {
        return 0;
    }

    /**
     * Create a new Line Follower sensor attached to a port
     * @param index the port index, eg 1, 2
     */
    //% blockId="bot_create_line_follower"
    //% block="create line follower at port %index"
    //% subcategory="LineFollower" weight=90 index.min=0 index.max=2
    export function createLineFollower(index: number): LineFollower {
        return new LineFollower(index);
    }

    /**
     * A LineFollower Sensor module, usually attached to a port
     */
    //%
    export class LineFollower {
        id: number;

        constructor(index: number) {
            this.id = lowLevelLineFollowerConstruct(index);
        }

        /**
         * Attaches code to run when the line follower occurs some event
         */
        //% blockId=bot_line_follower_on_event
        //% block="%this|on %event"
        //% blockGap=8 weight=82
        //% subcategory="LineFollower"
        onEvent(event: LineFollowerEvent, handler: () => void): void {
            control.onEvent(this.id, event, handler);
        }
    }

}
```
